ompi/communicator/comm.c

```diff
@@ -1010,130 +1143,73 @@ int ompi_comm_split_type (ompi_communicator_t *comm, int split_type, int key,
         return OMPI_SUCCESS;
     }
 
-    /* TODO: Make this better...
-     *
-     * See Example 7.4 in the MPI 4.0 standard for example usage.
-     *
-     * Stage 0: Recognized, but not implemented.
-     * Stage 1: Do better than that
-     */
-    if (MPI_COMM_TYPE_HW_UNGUIDED == global_orig_split_type) {
-        *newcomm = MPI_COMM_NULL;
-        return OMPI_SUCCESS;
-    }
-
-    /* Step 2: Build potential communicator groups. If any ranks will not be part of
-     * the ultimate communicator we will drop them later. This saves doing an extra
-     * allgather on the whole communicator. By using ompi_comm_split() later only
-     * if needed we 1) optimized the common case (no MPI_UNDEFINED and no reorder),
-     * and 2) limit the allgather to a smaller set of peers in the uncommon case. */
-    /* --------------------------------------------------------- */
-
-    /* allowed splitting types:
-       CLUSTER
-       CU
-       HOST
-       BOARD
-       NODE
-       NUMA
-       SOCKET
-       L3CACHE
-       L2CACHE
-       L1CACHE
-       CORE
-       HWTHREAD
-       Even though HWTHREAD/CORE etc. is overkill they are here for consistency.
-       They will most likely return a communicator which is equal to MPI_COMM_SELF
-       Unless oversubscribing.
-    */
-
-    /* how many ranks are potentially participating and on my node? */
-    rc = ompi_comm_split_type_get_part (comm->c_local_group, global_split_type, &lranks, &my_size);
-    if (OPAL_UNLIKELY(OMPI_SUCCESS != rc)) {
-        return rc;
-    }
-
-    /* Step 3: determine all the information for the remote group */
-    /* --------------------------------------------------------- */
-    if (inter) {
-        rc = ompi_comm_split_type_get_part (comm->c_remote_group, global_split_type, &rranks, &my_rsize);
-        if (OPAL_UNLIKELY(OMPI_SUCCESS != rc)) {
-            free (lranks);
-            return rc;
+    /* MPI_COMM_TYPE_HW_GUIDED: Check if 'value' the same at all ranks */
+    if (tmp[4] != -tmp[5]) {
+        if (0 == ompi_comm_rank(comm)) {
+            opal_output(0, "Error: Mismatched info values for MPI_COMM_TYPE_HW_GUIDED");
         }
+        return OMPI_ERR_BAD_PARAM;
     }
```

Comment:
This error check is now handled above in the `if (tmp[0] != -tmp[1] || tmp[4] != -tmp[5] || inter)` block. So can be removed.